sycl/test-e2e/Basic/vector/int-convert.cpp

```diff
@@ -0,0 +1,148 @@
+template <> std::string to_string<unsigned short>() { return "unsigned short"; }
+template <> std::string to_string<unsigned int>() { return "unsigned int"; }
+template <> std::string to_string<unsigned long>() { return "unsigned long"; }
+template <> std::string to_string<unsigned long long>() {
+  return "unsigned long long";
+}
+template <> std::string to_string<bool>() { return "bool"; }
+
+#define DEBUG_PRINT(x) std::cout << x << std::endl;
+
+#else
+#define DEBUG_PRINT(x)
+#endif
+
+template <typename T>
+void check_vectors_equal(sycl::vec<T, 4> a, sycl::vec<T, 4> b) {
+  bool all_good =
+      a.x() == b.x() && a.y() == b.y() && a.z() == b.z() && a.w() == b.w();
+  if (!all_good) {
+    DEBUG_PRINT("host and device results mismatch:");
+    DEBUG_PRINT(
+        "\t{" << static_cast<int>(a.x()) << ", " << static_cast<int>(a.y())
+              << ", " << static_cast<int>(a.z()) << ", "
+              << static_cast<int>(a.w()) << "} vs {" << static_cast<int>(b.x())
+              << ", " << static_cast<int>(b.y()) << ", "
+              << static_cast<int>(b.z()) << ", " << static_cast<int>(b.w())
+              << "}");
+  }
+  assert(all_good);
+}
+
+template <typename From, typename To> void check_convert() {
+  DEBUG_PRINT("checking vec<" << to_string<From>() << ", 4>::convert<"
+                              << to_string<To>() << ">()");
+
+  sycl::vec<From, 4> input;
+  if constexpr (std::is_signed_v<From>) {
+    input = sycl::vec<From, 4>{static_cast<From>(37), static_cast<From>(0),
+                               static_cast<From>(-11), static_cast<From>(13)};
+  } else {
+    input = sycl::vec<From, 4>{static_cast<From>(37), static_cast<From>(0),
+                               static_cast<From>(11), static_cast<From>(13)};
+  }
+
+  sycl::vec<To, 4> hostResult = input.template convert<To>();
+
+  sycl::buffer<sycl::vec<To, 4>> buf(sycl::range{1});
+  sycl::queue q;
+  q.submit([&](sycl::handler &cgh) {
+     sycl::accessor acc(buf, cgh);
+     cgh.single_task([=]() { acc[0] = input.template convert<To>(); });
+   }).wait();
+
+  auto acc = buf.get_host_access();
+  auto deviceResult = acc[0];
+
+  // Host and device results must match.
+  check_vectors_equal(hostResult, deviceResult);
+
+  // And they should match with a reference, which is for integer conversions
+  // can be computed with a simple static_cast.
+  // Strictly speaking, integer conversions are underspecified in the SYCL 2020
+  // spec, but `static_cast` implementation matches SYCL-CTS, so we will leave
+  // it here for now as well.
+  // See https://github.com/KhronosGroup/SYCL-Docs/issues/492
+  assert(deviceResult.x() == static_cast<To>(input.x()));
+  assert(deviceResult.y() == static_cast<To>(input.y()));
+  assert(deviceResult.z() == static_cast<To>(input.z()));
+  assert(deviceResult.w() == static_cast<To>(input.w()));
+}
+
+template <class T>
+constexpr auto has_unsigned_v =
```

Comment:
`is_supported_v` instead? Or `has_unsigned_convert_v`? I feel like `has_unsigned_v` might be a bit innacurate.

Reply:
Technically, we are checking if a type `T` has `unsigned` variant, so the name seems accurate to me.

This helper is used to shorten the test: as you can see, in `main` and `check_signed_unsigned_convert_to` we only list signed versions of types, but the test actually checks `unsigned -> signed`, `signed -> unsigned` and `unsigned -> unsigned` conversions for `To` and `From` types if they have unsigned variants.